### PR TITLE
Add spans to errors and clean up a lot of compiler code

### PR DESCRIPTION
**Pull request by LucasPickering**
- Added a new `Span` type to hold span metadata
- Reworked the external compiler API so you construct a `Compiler` and call methods on it, instead of just having public functions
- Reworked the errors a lot to hold span and source data
  - Each individual error holds the error type, the span of the code that triggered it, and the source located at that span (just a small copy of the full source code).
  - The parse errors still aren't good (I actually prob made them worse in this PR), I'm gonna work on that next.
- Fixed a bunch of broken doc links
  - Turns out the `<Target>` syntax doesn't work, idk why I thought it did. I changed those all to `[Target]` instead.
- Cleaned up a few feature flags that we aren't using anymore

#### Source

```
READ RX1
PUSH RX1 S0
POP S0 RLI
LBL:
LBL:
LBL:
JMP LBL1
```

#### Output

```
Error on line 5: Duplicate decalaration of label `LBL:`, originally defined on line 4
    |
  4 | LBL:
  5 | LBL:
    | ^^^^
  6 | LBL:
    |

Error on line 6: Duplicate decalaration of label `LBL:`, originally defined on line 4
    |
  5 | LBL:
  6 | LBL:
    | ^^^^
  7 | JMP LBL1
    |

Error on line 1: Invalid reference to register `RX1`
    |
  1 | READ RX1
    |      ^^^
  2 | PUSH RX1 S0
    |

Error on line 2: Invalid reference to register `RX1`
    |
  1 | READ RX1
  2 | PUSH RX1 S0
    |      ^^^
  3 | POP S0 RLI
    |

Error on line 2: Invalid reference to stack `S0`
    |
  1 | READ RX1
  2 | PUSH RX1 S0
    |          ^^
  3 | POP S0 RLI
    |

Error on line 3: Invalid reference to stack `S0`
    |
  2 | PUSH RX1 S0
  3 | POP S0 RLI
    |     ^^
  4 | LBL:
    |

Error on line 7: Invalid reference to label `LBL1`
    |
  6 | LBL:
  7 | JMP LBL1
    |     ^^^^
    |
```

#### Source

```
START:
READ RX0
JMP START
```

#### Output

```
Error on line 2: Read attempted while input is empty
    |
  1 | START:
  2 | READ RX0
    | ^^^^^^^^
  3 | JMP START
    |
```

### CODE REVIEW

File: core/src/ast.rs [hunk 1]
```diff
@@ -157,39 +133,42 @@ pub enum Jump<T> {
     Jgz(Node<ValueSource<T>, T>),
 }
 
-/// All types unique to the source AST live here
+/// All types unique to the source AST live here.
```
Comment (JRMurr) [inline core/src/ast.rs]: lowkey dont like the period but im cool if you have a reason for it

File: core/src/parse.rs [hunk 2]
```diff
@@ -324,35 +327,54 @@ fn line(input: RawSpan) -> ParseResult<'_, Option<SpanNode<Statement<Span>>>> {
...
+    input: &str,
+) -> Result<Program<Span>, Vec<SourceErrorWrapper<CompileError>>> {
     match Program::parse(RawSpan::new(input)) {
         Ok((_, program)) => Ok(program),
         Err(nom::Err::Error(e)) | Err(nom::Err::Failure(e)) => {
-            match e.errors.as_slice() {
-                [(substring, VerboseErrorKind::Nom(ErrorKind::Eof)), ..] => {
-                    // If the error is EOF that means there was remaining
-                    // input that was not parsed
-                    // so they put in a bad keyword
```
Comment (JRMurr) [inline core/src/parse.rs]: Is this comment not relevant anymore?

File: core/src/ast.rs (hunk 1 above)
Comment (LucasPickering) [reply core/src/ast.rs]: looked weird without it in the doc browser

File: core/src/parse.rs (hunk 2 above)
Comment (LucasPickering) [reply core/src/parse.rs]: Technically it still is now, but I got rid of it while I was trying to port this code. I'm planning on cleaning up the parsing errors soon, and I'm hoping we don't need to handle this case like this anymore.

**Pull request merged**